### ISSUE TITLE
Right click Layout preview to set center point

### ISSUE THREAD

**Issue by hayesp2527** [smeighan/xLights#3915]
I would like to see the ability to perform the following:

- Select a group in the layout tab
- Right click on the preview window and select a context menu for center point
- Options to set X, Y or both

This would allow quickly setting the center rather than the current type a number, see where it is, type another see where it is, etc
This is likely much easier in 2d view but if it could be done in 3d view that would be a bonus

**Comment by dartunder** [comment]
One quirk of using an offset for a group center point is that if you add a model to the group, the center point changes.
Is there a way to set an absolute position that would not change based on when items are added or removed from a group?

The "position" of the group ( if you select a group )  in the status bar is not expressed in absolute terms either.  It shows the group matrix's center point but not the actual position.

**Comment by AzGilrock** [comment]
I’ve spent a lot of time on trying to make it so the center point wouldn’t change when you add new models…it’s hard as fuck.  The center position if currently defined as an offset from the center of mass so anytime the models in the group change the center of mass changes.   I tried to capture the absolute position and restore it and it was close to working but then I found a couple cases where it was screwed up and I punted

**Comment by AzGilrock** [comment]
It’s a case of where we did the wrong thing and would need to screw up everyone’s center pints to fix it

**Comment by dartunder** [comment]
Yeah, I glanced at it, and it's not a trivial change; I think it requires a new property or checkbox to make it absolute.  One thought was to lock the position relative to a single specified prop, but not sure.

**Comment by AzGilrock** [comment]
I think the problem is a model group doesn’t really have an absolute position in the layout.

**Comment by merryoncherry** [comment]
It would be very interesting and much more useful to be able to center it on a prop.  If no prop specified, then stick to the old way... that way it wouldn't break everything.

**Comment by dartunder** [comment]
I have an initial start on some backward compatible code for "relative to a single specified prop", but I needed to focus on getting my display ready.  But don't let that stop you if someone else feels they know how to implement this.

Step one would be to allow the selection of either "group center" or "first item" and then maybe enable the selection of a specific prop.  Conceptually, I believe we need to calculate the selected items' effective offset from the group center, then add the manually specified offset and use the combined value as the previous offset for rendering.

This would still cause the center location to move if a model in the group is changed or added but if the manual offset is zero the group center would follow the selected model.

@AzGilrock said he had worked on some related code.... perhaps that code would help with the calculations needed?

**Comment by AzGilrock** [comment]
Well the issue with all these ideas is they run onto the same hurdle I had trouble with and that’s what happens when a model gets added to the group.  It’s not easy to fix the center point because that calculation is triggered later on.  Also imagine a model gets deleted that was a member of 6 models groups.  There’s not a mechanism to trigger the recalculation of all the center points for all those model groups.   Not saying it can’t be done but there are several scenarios that need to be tested with any change and most folks can’t think of everything that should be tested.

**Comment by dartunder** [comment]
I hear you; I only have a start of a concept anyway.    I expect the offset of the selected model from the group center would have to be calculated and added to the manual offset.  Calculating it on startup is one thing, but as you say, knowing when other changes happen is another thing.

**Comment by atticusm** [comment]
I would like to add the request to be able to lock the center point. At least now I know if I don't adjust the group the center won't change. I thought *any* change would cause the center to go off.  As long as I don't move any props belonging to that group it will stay centered.

**Comment by AzGilrock** [comment]
> I would like to add the request to be able to lock the center point. At least now I know if I don't adjust the group the center won't change. I thought _any_ change would cause the center to go off. As long as I don't move any props belonging to that group it will stay centered.

Sorry but if it was so simple to lock it we would have already done it
